### ISSUE TITLE
Data Race in entry.go:128

### ISSUE THREAD

**Issue by cosminilie** [sirupsen/logrus#295]
There seems to be a data race between github.com/Sirupsen/logrus.(*Entry).Info() at line 128  and the github.com/Sirupsen/logrus.SetLevel() method. I guess this applies to all methods on the Entry (Info, Warn, Error, Debug) at the if statement "if entry.Logger.Level >= InfoLevel {....."

This happens to me, because i want to update the loglevel externally (http endpoint).

Ex:

``` golang
go func() {
        for {
            level := <-logLevel
            log.SetLevel(level)
        }
}()

//Business domain
    go func() {
        log.WithFields(log.Fields{
            "File":   "main.go",
            "Action": "BussinessLogicInit",
        }).Info("Started Main Bussiness Unit")
   ......
}()
```

**Comment by Sirupsen** [comment]
Patch very welcome :+1:

**Comment by cosminilie** [comment]
Will try to send a patch. Thank you

**Comment by ash2k** [comment]
Same thing about all the other fields - `Formatter`, `Hooks`, `Out`.
e.g. `log.SetFormatter(...)` accesses field `Formatter` of the global instance while holding the mutex, but code in `entry.go` is reading from the field not holding the mutex.

**Comment by rand99** [comment]
yep, server down

```
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 97 [running]:
panic(0x739d80, 0xc848524040)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
sync.(*WaitGroup).Wait(0xc8202ee968)
	/usr/local/go/src/sync/waitgroup.go:129 +0x114
vendor/github.com/gemnasium/logrus-graylog-hook.(*GraylogHook).Flush(0xc8202ee940)
	vendor/github.com/gemnasium/logrus-graylog-hook/graylog_hook.go:91 +0x2d
vendor/github.com/gemnasium/logrus-graylog-hook.(*GraylogHook).Fire(0xc8202ee940, 0xc8470a2f00, 0x0, 0x0)
	vendor/github.com/gemnasium/logrus-graylog-hook/graylog_hook.go:84 +0xf2
vendor/github.com/Sirupsen/logrus.LevelHooks.Fire(0xc8202f4420, 0xc816a31a02, 0xc8470a2f00, 0x0, 0x0)
	/vendor/github.com/Sirupsen/logrus/hooks.go:28 +0x106
github.com/Sirupsen/logrus.Entry.log(0xc8202ee8c0, 0xc8484b5bc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1300002, 0xc84851c360, ...)
vendor/github.com/Sirupsen/logrus/entry.go:88 +0x119
vendor/github.com/Sirupsen/logrus.(*Entry).Error(0xc8470a2ec0, 0xc846b2d9a8, 0x1, 0x1)
vendor/github.com/Sirupsen/logrus/entry.go:145 +0x8c

```

**Comment by ash2k** [comment]
Ideally logging framework should not be thread-safe by default and provide means to wrap logger instance with a synchronizing wrapper. Take a look at `math/rand` package and how global `rand` is using a wrapped `Source` to add locking with mutex.
Global instance may be thread safe, but IMHO it should not be necessary unless user wants to do logger configuration concurrently with a running program. Most users probably configure logger at startup and do not touch it after so locking is pure overhead.
It may not be possible to retrofit such change without breaking the current API of course.

**Comment by talbright** [comment]
+1 

We need to make sure the library is thread safe...

**Comment by rand99** [comment]
i ended up using this (github.com/robertkowalski/graylog-golang) to send a marshalled json to the logserver and thats it :), survives the loadtest, keep it simple and stupid

**Comment by klizhentas** [comment]
@Sirupsen can you please take a look:

https://github.com/Sirupsen/logrus/pull/374

**Comment by sherryCP** [comment]
I just can't understand that

 **serialized, err := entry.Logger.Formatter.Format(entry)**

 is before

 **entry.Logger.mu.Lock()**

in the following function 

func (entry *Entry) write() {
	serialized, err := entry.Logger.Formatter.Format(entry)
	entry.Logger.mu.Lock()
	defer entry.Logger.mu.Unlock()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to obtain reader, %v\n", err)
	} else {
		_, err = entry.Logger.Out.Write(serialized)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to write to log, %v\n", err)
		}
	}
}

**Comment by nianglao** [comment]
@sherryCP Maybe your problem relates to https://github.com/sirupsen/logrus/issues?utf8=%E2%9C%93&q=Concurrent+map 

Please paste more detail info about how to reproduce this case.

**Comment by dgsb** [comment]
Fixed in #695
